### PR TITLE
Add CBMC Mini Projects list

### PR DESCRIPTION
**Pull request by karkhaz**
This list is intended to be a central repository of small, focussed CBMC
'nice-to-have' features that would be suitable for first-time CBMC
contributors to get their feet wet with. Projects on this list should
have the following features:

- Small (e.g. <200 LOC) or easy to implement

  - This is so that time-constrained contributors (like interns or
    students) can use these projects to ramp up before starting their
    main CBMC work

- Focussed, touching only one part of the codebase

  - This is so that people wanting to learn about one aspect of CBMC
    (solver, front-end, symex) can pick a project that involves that
    part of the codebase.

- Genuinely useful to somebody

  - This is so that somebody will be keen to quickly review and merge
    the code, and guide the new contributor in polishing the PR. This
    leads to a more rewarding experience for the contributor.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0944cdc3d7b4de63665080ecafb3fb518e6e0fb8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98560297
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: MINI-PROJECTS.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+**Background**: CBMC often doesn't know which function a function
+pointer points to. If a function is called through a function pointer,
+CBMC currently assumes that *any* function that has the same signature
+could be called. This results in a combinatorial explosion of paths when
+CBMC is exploring the program. It would be useful to have an annotation
+that tells CBMC exactly what function a pointer is expected to point to.
+
+
+### Generalize loop-unwinding bounds to cover multiple loops
```
Comment (danpoe) [inline MINI-PROJECTS.md:62]: Can we add a bit more information here to say that it should be done in cbmc and in the goto symex part? It's probably a bit hard to figure out where to start from the current description.

**Review by danpoe — COMMENTED** [review]
That looks very useful! I like the suggested projects as well and hope we'll add more in the future.

File: MINI-PROJECTS.md (hunk 1 above)
Comment (karkhaz) [reply MINI-PROJECTS.md:62]: Thanks, I've added a few hopefully helpful hints.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5764b81e81f6cdfea14f2716ea96437932d483fd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98656572
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by martin-cs** [comment]
Years ago (on the old Wiki) I wrote a bunch of these:

https://web.archive.org/web/20150530010030/http://www.cprover.org/wiki/doku.php?id=starter_projects

I don't know how many of them are still relevant / interesting.

**Review by chrisr-diffblue — APPROVED** [review]

**Comment by karkhaz** [comment]
Thanks! @martin-cs please feel free to add any that are still relevant? It looks like some of them were already assigned to people, and a couple of them look complete.

Also, many of those projects look interesting and useful, but are larger than the ones I envisioned for that list. I do think it's worth keeping a record of larger starter projects though. I now wonder whether this would be better maintained as a page under the official documentation, split into subsections (<= 2 weeks, <= 6 weeks, <= 3 month projects)?

I don't mind either way, perhaps we could merge this and then add some longer projects in a separate PR.

**Pull request merged**